libvisual-plugins/plugins/actor/dancingparticles/font.cpp

```diff
@@ -0,0 +1,229 @@
+  int numPart = 0;
+  int scoreToGo = 1;
+  for(int y = 0; y < height ; y++)
+    {
+      for(int x = 0; x < width ; x++)
+        {
+          unsigned char c = pixels[y*pitch+x];
+          int a = 0;
+          lastscore = curscore;
+          curscore +=c;
+          if(c!=0)
+            {
+              while(lastscore < scoreToGo  && curscore >= scoreToGo)
+                {
+                  Centers[numPart++]=FloatPoint(x*4-width*2,height*8-y*16+a,0);
+                  scoreToGo= (numPart*maxscore)/numCenters;
+                  a+=4;
+                }
+            }
+        }
+    }
+}
+
+VisVideo *rasteriseText(FT_Face face, const std::string &text)
+{
+  // This is a very simple FT2 text rasteriser that supports languages with a one-one character to glyph mapping.
+  // Anything else would require a text shaping engine like Harfbuzz.
+
+  FT_Error error;
+
+  bool useKerning = FT_HAS_KERNING(face);
+
+  std::size_t charCount = text.length();
+
+  std::vector<Glyph> glyphs;
+  glyphs.reserve(charCount);
+
+  auto ftGlyphSlot = face->glyph;
+
+  // Convert text string into a series of glyphs.
+
+  int penX = 0;
+  int penY = 0;
+  FT_UInt prevFtGlyphIndex = 0;
+
+  for(auto text_char : text)
+    {
+      auto ftGlyphIndex = FT_Get_Char_Index(face, text_char);
+
+      error = FT_Load_Glyph(face, ftGlyphIndex, FT_LOAD_DEFAULT);
+      if(error)
+        continue;
+
+      FT_Glyph ftGlyph = nullptr;
+
+      error = FT_Get_Glyph(ftGlyphSlot, &ftGlyph);
+      if(error)
+        continue;
+
+      if(useKerning && prevFtGlyphIndex && ftGlyphIndex)
+        {
+          FT_Vector delta;
+          FT_Get_Kerning(face, prevFtGlyphIndex, ftGlyphIndex, FT_KERNING_DEFAULT, &delta);
+          penX += delta.x >> 6;
+        }
+
+      glyphs.push_back({ FT_Vector { penX, penY }, ftGlyph });
+
+      penX += ftGlyphSlot->advance.x >> 6;
+
+      prevFtGlyphIndex = ftGlyphIndex;
+    }
+
+  // Calculate bounding box of text rasterization.
+  //
+  // NOTE: According to the FT2 tutorial, FT_Glyph_Get_CBox() does not account for anti-aliasing.
+  // However, the returned bounds appear accurate for our purpose.
+
+  FT_BBox textBBox
+    {
+      std::numeric_limits<int>::max(),
+      std::numeric_limits<int>::max(),
+      std::numeric_limits<int>::min(),
+      std::numeric_limits<int>::min()
+    };
+
+  for(auto const& glyph : glyphs)
+    {
+      FT_BBox glyphBBox { 0, 0, 0, 0 };
+      FT_Glyph_Get_CBox(glyph.ftGlyph, ft_glyph_bbox_pixels, &glyphBBox);
+
+      glyphBBox.xMin += glyph.pos.x;
+      glyphBBox.yMin += glyph.pos.y;
+      glyphBBox.xMax += glyph.pos.x;
+      glyphBBox.yMax += glyph.pos.y;
+
+      textBBox.xMin = std::min(textBBox.xMin, glyphBBox.xMin);
+      textBBox.yMin = std::min(textBBox.yMin, glyphBBox.yMin);
+      textBBox.xMax = std::max(textBBox.xMax, glyphBBox.xMax);
+      textBBox.yMax = std::max(textBBox.yMax, glyphBBox.yMax);
+    }
+
+  if(textBBox.xMin > textBBox.xMax)
+    return nullptr;
```

Comment:
Do we need to loop `FT_Done_Glyph` here?

Reply:
Yes, we do. I will work it into the Glyph class destructor to avoid manual cleanups.